scylla/src/transport/session.rs

```diff
@@ -1321,12 +1351,16 @@ impl Session {
                         .instrument(span.clone())
                         .await;
 
+                let elapsed = query_start.elapsed();
+                if self.should_consider_query_for_latency_measurements(&query_result) {
+                    let mut average_latency_guard = node.average_latency.write().unwrap();
```

Comment:
We could create a channel on which latency measurements would be sent.
`NodeWorker` would collect these measurements and periodically update atomic with current latency.

This would offload the cost of doing `thread_local` super optimized stuff onto the channel library,  something like crossbeam.

Reply:
We could create a channel on which latency measurements would be sent.
`NodeWorker` would collect these measurements and periodically update atomic with current latency.

This would offload the cost of doing `thread_local` super optimized stuff onto the channel library,  something like crossbeam.

Reply:
I consider this problem as open, and I have no personal preference nor conviction towards any particular solution. It seems though to be the last (yet important) issue with the PR, so I encourage you to further discussion.

Reply:
I'm not sure if I like the channel solution because either the channel has to be bounded and we might have to either drop measurements/block until the channel is cleared, or the channel is unbounded and its memory usage is theoretically unbounded.

IMO for now we can proceed with the current solution - if the original implementation performs similar locks then we shouldn't be much worse performance wise. In the future, if we have more data and the locks turn out to be a bottleneck, we might consider optimizing it. Ultimately, all those problems are opt-in and require the latency-aware policy to be turned on.

@psarna @cvybhu what do you think?

Reply:
Let's release fast and optimize on demand.